### ISSUE TITLE
[feature request] [ux] Frontend methods for fused mul+add+dtype convert

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#106624]
### 🚀 The feature, motivation and pitch

This should be well-compiled, but would be nice to have as a frontend method.

A prototypical example is conversion of float32 image to uint8 image:

```python
import torch
a = torch.rand(10)
b = torch.empty_like(a, dtype = torch.uint8)
f32_to_u8 = torch.mul(a, 255, out = b)
# RuntimeError: result type Float can't be cast to the desired output type Byte
```

Might be that these methods already exist in the form of helpers for quantization, but would be nice to expose them for wider usage.

OpenCV https://docs.opencv.org/4.x/d3/d63/classcv_1_1Mat.html#adf88c60c5b4980e05bb556080916978b has a similar method `convertTo` which multiplies by a scalar, than adds another scalar and then casts dtype. OpenCV also does saturate cast functionality (to make sure that uint8 doesn't overflow too badly) - might be useful for PyTorch too, but saturation might be done prior to actual conversion by `clamp_`.

Related: `addcmul` generalization discussion: https://github.com/pytorch/pytorch/issues/104849

This might also be too niche,

**Comment by vadimkantorov** [comment]
Unfortunately, hijacking `torch.div(..., rounding_mode = 'trunc')` for float32->uint8 conversion does not work:
```python
import torch

a = torch.rand(10, dtype = torch.float32)
b = torch.empty_like(a, dtype = torch.uint8)

torch.div(a, 1/255.0, rounding_mode = 'trunc', out = b)
# RuntimeError: result type Float can't be cast to the desired output type Byte

#a.div(1/255.0, rounding_mode = 'trunc', out = b)
#TypeError: div() received an invalid combination of arguments - got (float, out=Tensor, rounding_mode=str), but expected one of:
# * (Tensor other)
# * (Tensor other, *, str rounding_mode)
#      didn't match because some of the keywords were incorrect: out
# * (Number other, *, str rounding_mode)
#      didn't match because some of the keywords were incorrect: out
```

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/42959

**Comment by vadimkantorov** [comment]
Maybe! As long as the ops don't aggresively guard against downcasting.

+ need to check that torch.add/torch.mul works well, doesn't upcast inputs too much, and produces nice fused code

**Comment by vadimkantorov** [comment]
Btw currently, torch.sum supports already directly a `dtype=` argument (but is accumulation still done in int32 in all cases?):
```
>>> torch.tensor([255, 255], dtype = torch.uint8).sum().dtype
torch.int64
>>> torch.tensor([255, 255], dtype = torch.uint8).sum(dtype = torch.uint8).dtype
torch.uint8
```

One problem is that it does upcast its inputs: https://github.com/pytorch/pytorch/issues/55366

But in general I guess it's useful to separate various compute_dtype and out_dtype

Another issue is `clamp` before storing the output (to emulate saturate_cast). One way might be simply having recipes in docs that are tested to produce good fusion code.

But as I found in https://github.com/pytorch/pytorch/issues/106614, torch.compile is non-local in a sense, as dynamo options seem global. IMO this is quite a bad design decision as the codegen'd code can change depending on some spooky code operating at a distance in some dependency.

**Comment by albanD** [comment]
> But in general I guess it's useful to separate various compute_dtype and out_dtype

Yes we definitely want to separate these two. There are ops such as batchnorm that require high precision compute even for small dtype input/outputs.

cc @jansel for more general dynamo setting discussion

**Comment by vadimkantorov** [comment]
Maybe related on saturation :) https://github.com/pytorch/pytorch/issues/106780

**Comment by vadimkantorov** [comment]
@alband Funny saturation/casts quirks:
```python
import torch

x = torch.tensor([True], dtype = torch.bool)

print(torch.full_like(x, float('nan')))
# tensor([True])
print(torch.full_like(x, float('inf')))
# tensor([True])
print(torch.full_like(x, float('-inf')))
# tensor([True])
print(torch.full_like(x, float('-0')))
# tensor([False])
```

**Comment by albanD** [comment]
Isn't it expected that every non-zero value is True?

**Comment by vadimkantorov** [comment]
It is, but it's a bit surprising for NaN and also a bit surprising how full_like infers dtype: from the fill_value or from the passed tensor

**Comment by albanD** [comment]
> bit surprising for NaN (I would maybe expect it map to False) 

Converting a NaN to any dtype that doesn't have a NaN representation is always going to be weird I think.
At best, we could make this an error if you think that it is better.

> a bit surprising how full_like infers dtype: from the fill_value or from the passed tensor

How so? Just like everything else, python numbers/scalars do no participate in type promotion.

**Comment by vadimkantorov** [comment]
> How so? Just like everything else, python numbers/scalars do no participate in type promotion.

I guess I was confused by type inference from `torch.tensor(5)` vs. `torch.tensor(5.0)` :)

And maybe it's worth a mention specifically for torch.full_like as it's where there potentially a conflict between scalar dtype and tensor dtype

**Comment by vadimkantorov** [comment]
It appears that PyTorch also does not support `out=` for `.to`/`.astype` :(

**Comment by vadimkantorov** [comment]
It appears that one can use `copy_` for such casts, but it's not fully clear how saturation works. I can't understand where `32536` is coming from, given that int16 ranges from -32768 to 32767:

```python
import torch

a = torch.tensor([1,2,3], dtype = torch.int16)
print(a) # tensor([1, 2, 3], dtype=torch.int16)

a.copy_(torch.tensor([4.0, 5.0, 6.0], dtype = torch.float32))
print(a) # tensor([4, 5, 6], dtype=torch.int16)

a.copy_(torch.tensor([-32769.0, 0.0, 32769.0], dtype = torch.float32))
print(a) # tensor([ 32767,      0, -32767], dtype=torch.int16)

a.copy_(torch.tensor([-33000.0, 0.0, 33000.0], dtype = torch.float32))
print(a) # tensor([ 32536,      0, -32536], dtype=torch.int16)
```

**Comment by vadimkantorov** [comment]
Related usecase:
- https://github.com/pytorch/vision/issues/8669#issuecomment-2387963631

**Comment by vadimkantorov** [comment]
in OpenCV this exists as `cv2.scaleAdd`:
- https://docs.opencv.org/4.x/d2/de8/group__core__array.html#ga9e0845db4135f55dcf20227402f00d98
